src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,426 @@
+            switch (currentNode.Kind())
+            {
+                case SyntaxKind.DoStatement:
+                    var dostatement = (DoStatementSyntax)currentNode;
+                    return dostatement.CloseParenToken.IsMissing;
+                case SyntaxKind.ForStatement:
+                    var forStatement = (ForStatementSyntax)currentNode;
+                    return forStatement.CloseParenToken.IsMissing;
+                default:
+                    return false;
+            }
+        }
+
+        /// <summary>
+        /// Determines if a syntax node includes all required closing delimiters.
+        /// </summary>
+        /// <remarks>
+        /// <para>Some syntax nodes, such as parenthesized expressions, require a matching closing delimiter to end the
+        /// syntax node. If this node is omitted from the source code, the parser will automatically insert a zero-width
+        /// "missing" closing delimiter token to produce a valid syntax tree. This method determines if required closing
+        /// delimiters are present in the original source.</para>
+        /// </remarks>
+        /// <param name="currentNode"></param>
+        /// <returns>
+        /// <list type="bullet">
+        /// <item><description><see langword="true"/> if <paramref name="currentNode"/> requires a closing delimiter and the closing delimiter is present in the source (i.e. not missing)</description></item>
+        /// <item><description><see langword="true"/> if <paramref name="currentNode"/> does not require a closing delimiter</description></item>
+        /// <item><description>otherwise, <see langword="false"/>.</description></item>
+        /// </list>
+        /// </returns>
+        private static bool ClosingDelimiterExistsIfNeeded(SyntaxNode currentNode)
+        {
+            switch (currentNode.Kind())
+            {
+                case SyntaxKind.ArgumentList:
+                    var argumentList = (ArgumentListSyntax)currentNode;
+                    return !argumentList.CloseParenToken.IsMissing;
+
+                case SyntaxKind.ParenthesizedExpression:
+                    var parenthesizedExpression = (ParenthesizedExpressionSyntax)currentNode;
+                    return !parenthesizedExpression.CloseParenToken.IsMissing;
+
+                case SyntaxKind.BracketedArgumentList:
+                    var bracketedArgumentList = (BracketedArgumentListSyntax)currentNode;
+                    return !bracketedArgumentList.CloseBracketToken.IsMissing;
+
+                case SyntaxKind.ObjectInitializerExpression:
+                    var initializerExpressionSyntax = (InitializerExpressionSyntax)currentNode;
+                    return !initializerExpressionSyntax.CloseBraceToken.IsMissing;
+
+                case SyntaxKind.ArrayRankSpecifier:
+                    var arrayRankSpecifierSyntax = (ArrayRankSpecifierSyntax)currentNode;
+                    return !arrayRankSpecifierSyntax.CloseBracketToken.IsMissing;
```

Comment:
a similar comment about how this can be simplified applies here.

Reply:
This is addressed in follow-up issue https://github.com/dotnet/roslyn/issues/32163